File: daily_challenge_leetcode/2300.js
// implementing own lower bound
// Tc = O(n log n + m log n)
// Sc = O(1) or O(m) [including the result array]

var successfulPairs = function (spells, potions, success) {
  const m = spells.length; // number of spells
  const n = potions.length; // number of potions

  potions.sort((a, b) => a - b); // sort potions in ascending order
  const maxPotion = potions[n - 1]; // largest potion value
  const answer = []; // stores number of successful pairs for each spell

  // 🔹 Helper function: binary search for first potion >= minPotion
  const lowerBound = (l, r, minPotion) => {
    let possibleIndex = -1;
    while (l <= r) {
      const mid = l + Math.floor((r - l) / 2); // midpoint

      if (potions[mid] >= minPotion) {
        possibleIndex = mid; // potential valid position
        r = mid - 1; // continue searching left for earlier valid potion
      } else {
        l = mid + 1; // search right
      }
    }
    return possibleIndex; // returns first valid index, or -1 if not found
  };

  // 🔹 For each spell, count how many potions form a successful pair
  for (let i = 0; i < m; i++) {
    const spell = spells[i];

    // minimum potion strength needed to reach or exceed success
    const minPotion = Math.ceil(success / spell);

    // if even the strongest potion is too weak, add 0
    if (minPotion > maxPotion) {
      answer.push(0);
      continue;
    }

    // binary search for the first potion >= minPotion
    const index = lowerBound(0, n - 1, minPotion);

    // if index === -1, no valid potions; otherwise count all from index → end
    answer.push(index === -1 ? 0 : n - index);
  }

  return answer;
};

File: daily_challenge_leetcode/2300.cpp
//Approach-1 : Using lower_bound STL
//T.C : O(n log n + m log n)
//S.C : O(1)
class Solution {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        int m = spells.size();
        
        int n = potions.size();
        
        //O(n * logn)
        sort(begin(potions), end(potions));
        
        int maxPotion = potions[n-1];
        
        vector<int> answer;
        
        //O(m * logm)
        for(int i = 0; i<m; i++) {
            
            int spell = spells[i];
            
            long long minPotion = ceil((1.0*success)/spell);
            
            if(minPotion > maxPotion) {
                answer.push_back(0);
                continue;
            }
            
            int index = lower_bound(begin(potions), end(potions), minPotion) - begin(potions);
            
            answer.push_back(n-index);
            
        }
        
        return answer;
    }
};

// implementing own lower bound
// Tc = O(n log n + m log n)
// Sc = O(1) or O(m) [including the result array]
class Solution {
public:
    int lowerBound(int l, int r, vector<int>& potions, int minPotion) {
        int possibleIndex = -1;
        int mid = 0;
        while (l <= r) {
            mid = l + (r - l) / 2;
            if (potions[mid] >= minPotion) {
                possibleIndex = mid;
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return possibleIndex;
    }
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions,
                                long long success) {
        int m = spells.size();

        int n = potions.size();

        sort(begin(potions), end(potions));

        int maxPotion = potions[n - 1];

        vector<int> answer;

        for (int i = 0; i < m; i++) {

            int spell = spells[i];

            long long minPotion = ceil((1.0 * success) / spell);

            if (minPotion > maxPotion) {
                answer.push_back(0);
                continue;
            }

            int index = lowerBound(0, n - 1, potions, minPotion);

            answer.push_back(n - index);
        }

        return answer;
    }
};
